File: libraries/ios/Sources/ChatUi/library/FlexChatUIBridge.swift
let kChatUiBaseUtl = "https://ui.chatkitty.com"

final class FlexChatUIBridge: @preconcurrency ChatUIBridge {
    
    private let component: FlexComponent
    
    @MainActor init(component: FlexComponent) {
        component.setBaseUrl(kChatUiBaseUtl)
        self.component = component
    }
    
    // MARK: ChatUIBridge
    
    @MainActor func initialize(options: InitializeOptions) {
        component.evalFlexFunc("initialize", sendData: options)
    }
    
    @MainActor func onChatUiConnected(_ onChatUiConnected: @escaping () -> Void) {
        component.setInterface("onChatUiConnected") { _ in
            onChatUiConnected()
        }
    }
    
    @MainActor func onChatMounted(_ onChatMounted: @escaping (ChatComponentContext) -> Void) {
        component.setInterface("onChatMounted") { (model: ChatComponentContext?) -> Any? in
            if let chatMountedContext = model {
                onChatMounted(chatMountedContext)
            }
            return nil
        }
    }
    
    @MainActor func onChatHeaderSelected(_ onChatHeaderSelected: @escaping (Channel) -> Void) {
        component.setInterface("onChatHeaderSelected") { (model: Channel?) -> Any? in
            if let channel = model {
                onChatHeaderSelected(channel)
            }
            return nil
        }
    }
    
    @MainActor func onChatMenuActionSelected(_ onChatMenuActionSelected: @escaping (MenuAction) -> Void) {
        component.setInterface("onChatMenuActionSelected") { (model: MenuAction?) -> Any? in
            if let menuAction = model {
                onChatMenuActionSelected(menuAction)
            }
            return nil
        }
    }
    
    @MainActor func onChatNotificationReceived(_ onChatNotificationReceived: @escaping (BaseNotification) -> Void) {
        component.setInterface("onChatNotificationReceived") { (model: BaseNotification?) -> Any? in
            if let notification = model {
                onChatNotificationReceived(notification)
            }
            return nil
        }
    }
    
    @MainActor func onPostMessage(_ onPostMessage: @escaping (ChatUIMessage) -> Void) {
        component.setInterface("postMessage")  { (model: ChatUIMessage?) -> Any? in
            if let message = model {
                onPostMessage(message)
            }
            return nil
        }
    }
}

File: libraries/android/chatkitty/src/main/java/com/chatkitty/library/FlexChatUIBridge.kt
package com.chatkitty.library

import app.dvkyun.flexhybridapp.FlexWebView
import com.chatkitty.api.ChatUIBridge
import com.chatkitty.models.ChatComponentContext
import com.chatkitty.models.ChatUIMessage
import com.chatkitty.models.InitializeOptions

class FlexChatUIBridge(private val flexWebView: FlexWebView) : ChatUIBridge {

    // ChatUIBridge implementation

    override fun initialize(options: InitializeOptions) {
        flexWebView.evalFlexFunc("initialize", options)
    }

    override fun onChatUiConnected(onChatUiConnected: () -> Unit) {
        flexWebView.setInterface("onChatUiConnected") { _ ->
            onChatUiConnected()
        }
    }

    override fun onChatMounted(onChatMounted: (ChatComponentContext) -> Unit) {
        flexWebView.typeInterface("onChatMounted", ChatComponentContext::class) { context ->
            onChatMounted(context)
        }
    }

    override fun onChatHeaderSelected(onChatHeaderSelected: () -> Unit) {
        flexWebView.typeInterface("onChatHeaderSelected", io.swagger.client.models.Channel::class) { context ->
            onChatHeaderSelected()
        }
    }

    override fun onChatMenuActionSelected(onChatMenuActionSelected: () -> Unit) {
        flexWebView.setInterface("onChatMenuActionSelected") { args ->
            onChatMenuActionSelected()
        }
    }

    override fun onChatNotificationReceived(onChatNotificationReceived: () -> Unit) {
        flexWebView.setInterface("onChatNotificationReceived") { args ->
            onChatNotificationReceived()
        }
    }

    override fun onPostMessage(onPostMessage: (ChatUIMessage) -> Unit) {
        flexWebView.typeInterface("postMessage", ChatUIMessage::class) { args ->
            onPostMessage(args)
        }
    }
}
